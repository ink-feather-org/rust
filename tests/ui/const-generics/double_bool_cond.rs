// [unified] run-pass
// revisions: unified normal
#![feature(generic_const_exprs)]
#![cfg_attr(unified, feature(impl_unified_exhaustive_const_traits))]
#![allow(incomplete_features)]

pub const fn bool_to_usize(b: bool) -> usize {
    b as usize
}

pub struct ConstTwo<const A: bool, const B: bool>;

impl Default for ConstTwo<true, true> {
  fn default() -> Self { Self }
}
impl Default for ConstTwo<true, false> {
  fn default() -> Self { Self }
}
impl Default for ConstTwo<false, true> {
  fn default() -> Self { Self }
}
impl Default for ConstTwo<false, false> {
  fn default() -> Self { Self }
}

#[derive(Default)]
pub struct Arg<const N: usize> where
  [(); bool_to_usize(N <= 5)]:, [(); bool_to_usize(N > 3)]:,
  {
    _c: ConstTwo<{ N <= 5 }, {N > 3}>,
    //[normal]~^ ERROR: the trait
}

fn main() {
  let _def = Arg::<2>::default();
}
