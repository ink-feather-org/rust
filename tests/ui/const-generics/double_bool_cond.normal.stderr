error[E0277]: the trait bound `ConstTwo<{ N <= 5 }, {N > 3}>: Default` is not satisfied
  --> $DIR/double_bool_cond.rs:30:5
   |
LL | #[derive(Default)]
   |          ------- in this derive macro expansion
...
LL |     _c: ConstTwo<{ N <= 5 }, {N > 3}>,
   |     ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ the trait `Default` is not implemented for `ConstTwo<{ N <= 5 }, {N > 3}>`
   |
   = help: the following other types implement trait `Default`:
             ConstTwo<false, false>
             ConstTwo<false, true>
             ConstTwo<true, false>
             ConstTwo<true, true>
   = note: this error originates in the derive macro `Default` (in Nightly builds, run with -Z macro-backtrace for more info)
help: consider annotating `ConstTwo<{ N <= 5 }, {N > 3}>` with `#[derive(Default)]`
   |
LL + #[derive(Default)]
LL | pub struct ConstTwo<const A: bool, const B: bool>;
   |

error: aborting due to previous error

For more information about this error, try `rustc --explain E0277`.
